### PR TITLE
DWARF type parsing prototype

### PR DESCRIPTION
**Pull request by HoundThe**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Description**

I've made this branch on top of unmerged PR because I wanted to do a little prototype, but I probably should've finished unit tests to merge it.. oh well..

I am starting to prototype out the type parsing out of dwarf information, right now I got a big ugly and naive solution, but I am getting there. I wanted to show the approach I started with, because I am really unsure with it, so I started implementing something to have something to kick off. I didn't know better way to discuss this than in PR so I made one.

Right now I am parsing the types to RAnalBaseType format and my plan is to make appropriate type setting API inside anal/type.c as I've done with type getting. 

I've kept the parsing methods inside dwarf.c as I think they belong there and exported the symbol so someone else will actually do the parsing.

changes are especially from L:336 to L:470 in dwarf.c

example now: (context, I can't parse type yet, so it's always `Type`)
```
Struct, name: human
 Members:
  Type : height
  Type : weight
  Type : name
Struct, name: donkey
 Members:
  Type : leg_count
  Type : face_length
  Type : name
  Type : owner
```

### CODE REVIEW

File: libr/bin/dwarf.c [hunk 1]
```diff
@@ -194,6 +298,176 @@ static const char *dwarf_langs[] = {
...
+
+			RBinDwarfDIE *curr_die = &info->comp_units[i].dies[j];
+			if (is_type_tag (curr_die->tag)) {
+				// now we control the flow, lets construct the structure
+				RAnalBaseType *base_type = R_NEW(RAnalBaseType);
+				RVector members;
+				// TODO add free function
+				r_vector_init (&members, sizeof (RAnalStructMember), struct_type_fini, NULL);
+
+				if (curr_die->tag == DW_TAG_structure_type) {
```
Comment (XVilka) [inline libr/bin/dwarf.c:401]: I suggest to extract that part in a separate function.

File: libr/bin/dwarf.c [hunk 2]
```diff
@@ -1294,19 +1596,20 @@ static void r_bin_dwarf_dump_debug_info(FILE *f, const RBinDwarfDebugInfo *inf)
 	}
 
 	for (i = 0; i < inf->length; i++) {
-		fprintf (f, "  Compilation Unit @ offset 0x%"PFMT64x":\n", inf->comp_units [i].offset);
-		fprintf (f, "   Length:        0x%x\n", inf->comp_units [i].hdr.length);
-		fprintf (f, "   Version:       %d\n", inf->comp_units [i].hdr.version);
-		fprintf (f, "   Abbrev Offset: 0x%x\n", inf->comp_units [i].hdr.abbrev_offset);
-		fprintf (f, "   Pointer Size:  %d\n", inf->comp_units [i].hdr.pointer_size);
+		fprintf (f, "\n");
```
Comment (XVilka) [inline libr/bin/dwarf.c:1599]: Can you please use `r_cons_printf` instead?

File: libr/bin/dwarf.c [hunk 3]
```diff
@@ -1346,293 +1647,495 @@ static const ut8 *r_bin_dwarf_parse_attr_value(const ut8 *obuf, int obuf_len,
...
 		break;
-	case DW_FORM_udata:
-		{
-			ut64 ndata = 0;
-			const ut8 *data = (const ut8*)&ndata;
-			buf = r_uleb128 (buf, R_MIN (sizeof (data), (size_t)(buf_end - buf)), &ndata);
-			memcpy (&value->encoding.data, data, sizeof (value->encoding.data));
-			value->encoding.str_struct.string = NULL;
+	case DW_FORM_ref_addr: // offset in .debug_info
+		// This is 4 or 8 bytes depending where it refers to
```
Comment (XVilka) [inline libr/bin/dwarf.c:1771]: Please fix the comment indentation.

File: libr/bin/dwarf.c [hunk 4]
```diff
@@ -1346,293 +1647,495 @@ static const ut8 *r_bin_dwarf_parse_attr_value(const ut8 *obuf, int obuf_len,
...
+			hdr->abbrev_offset = READ64 (buf);
+		} else {
+			hdr->abbrev_offset = READ32 (buf);
+		}
+		hdr->address_size = READ8 (buf);
+	}
+	hdr->header_size = buf - tmp; // header size excluding length field
+	return buf;
+}
+static int expand_info(RBinDwarfDebugInfo *info) {
```
Comment (XVilka) [inline libr/bin/dwarf.c:2041]: Missing space before.

File: test/unit/test_dwarf_info.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	mu_assert_eq (cu.length, 10, "Wrong attribute information");
+	mu_assert_eq (cu.offset, 0x0, "Wrong attribute information");
+
+	mu_assert_eq (cu.dies[0].abbrev_code, 1, "Wrong attribute information");
+	mu_assert_eq (cu.dies[0].length, 7, "Wrong attribute information");
+	mu_assert_eq (cu.dies[0].tag, DW_TAG_compile_unit, "Wrong attribute information");
+	mu_assert_eq (cu.dies[0].attr_values[0].attr_name, DW_AT_producer, "Wrong attribute information");
+	mu_assert_streq(cu.dies[0].attr_values[2].string.content, "main.c", "Wrong attribute information");
+
+	mu_assert_eq (cu.dies[1].abbrev_code, 2, "Wrong attribute information");
```
Comment (XVilka) [inline test/unit/test_dwarf_info.c:42]: Please compactify these with a macro.

File: libr/bin/dwarf.c (hunk 3 above)
Comment (HoundThe) [reply libr/bin/dwarf.c:1771]: yeah those fixes belong into the previous PR that this PR builds on, I'll merge the changes from there after

**Comment by XVilka** [comment]
Please rebase.

**Comment by HoundThe** [comment]
Closing, I made a new PR for this #17111

**Pull request closed without merging**